projects/com.oracle.truffle.llvm.asm.amd64/src/com/oracle/truffle/llvm/asm/amd64/AsmFactory.java

```diff
@@ -0,0 +1,199 @@
+        getArguments();
+        return new LLVMInlineAssemblyRootNode(null, frameDescriptor, statements.toArray(new LLVMNode[statements.size()]), arguments, result);
+    }
+
+    public void createBinaryOperation(String operation, String left, String right) {
+        FrameSlot leftSlot = frameDescriptor.findFrameSlot(left);
+        FrameSlot rightSlot = frameDescriptor.findFrameSlot(right);
+        LLVMI32Node leftNode = (leftSlot != null) ? LLVMI32ReadNodeGen.create(leftSlot) : getImmediateNode(left);
+        LLVMI32Node rightNode = LLVMI32ReadNodeGen.create(rightSlot);
+        this.result = LLVMI32ReadNodeGen.create(rightSlot);
+        switch (operation) {
+            case "addl":
+                this.statements.add(LLVMWriteI32NodeGen.create(LLVMAMD64AddlNodeGen.create(leftNode, rightNode), rightSlot));
+                break;
+            case "subl":
+                this.statements.add(LLVMWriteI32NodeGen.create(LLVMAMD64SublNodeGen.create(leftNode, rightNode), rightSlot));
+                break;
+            case "andl":
+                this.statements.add(LLVMWriteI32NodeGen.create(LLVMAMD64AndlNodeGen.create(leftNode, rightNode), rightSlot));
+                break;
+            case "orl":
+                this.statements.add(LLVMWriteI32NodeGen.create(LLVMAMD64OrlNodeGen.create(leftNode, rightNode), rightSlot));
+                break;
+            case "xorl":
+                this.statements.add(LLVMWriteI32NodeGen.create(LLVMAMD64XorlNodeGen.create(leftNode, rightNode), rightSlot));
+                break;
+            case "sall":
+                this.statements.add(LLVMWriteI32NodeGen.create(LLVMAMD64SallNodeGen.create(leftNode, rightNode), rightSlot));
+                break;
+            case "sarl":
+                this.statements.add(LLVMWriteI32NodeGen.create(LLVMAMD64SarlNodeGen.create(leftNode, rightNode), rightSlot));
+                break;
+            case "shll":
+                this.statements.add(LLVMWriteI32NodeGen.create(LLVMAMD64ShllNodeGen.create(leftNode, rightNode), rightSlot));
+                break;
+            case "shrl":
+                this.statements.add(LLVMWriteI32NodeGen.create(LLVMAMD64ShrlNodeGen.create(leftNode, rightNode), rightSlot));
+                break;
+            case "movl":
+                this.statements.add(LLVMWriteI32NodeGen.create(LLVMAMD64MovlNodeGen.create(leftNode, rightNode), rightSlot));
+                break;
+            default:
+                this.statements.add(new LLVMI32UnsupportedInlineAssemblerNode());
+        }
+    }
+
+    public void createUnaryOperation(String operation, String node) {
+        FrameSlot slot = frameDescriptor.findFrameSlot(node);
+        LLVMI32Node childNode = LLVMI32ReadNodeGen.create(slot);
+        this.result = LLVMI32ReadNodeGen.create(slot);
+        switch (operation) {
+            case "notl":
+                this.statements.add(LLVMWriteI32NodeGen.create(LLVMAMD64NotlNodeGen.create(childNode), slot));
```

Comment:
I think you can also avoid the redundancy of `this.statements.add(LLVMWriteI32NodeGen.create(...))` here.